src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -1004,15 +1034,29 @@ func (i *fetchTaggedResultsIter) acquire(ctx context.Context, idx int) (bool, er
 	if curPermit == nil || curPermit.QuotaRemaining() <= 0 {
 		if i.idx == idx {
 			// block acquiring if we need the block readers to fulfill the current fetch.
-			permit, err := i.blockPermits.Acquire(ctx)
+			permit, acquireResult, err := i.blockPermits.Acquire(ctx)
+			if acquireResult.Waited {
+				if err == nil && i.requireNoWait {
+					// Fail iteration if request requires no waiting.
+					return false, permits.ErrOperationWaitedOnRequireNoWait
+				}
+				i.seriesReadWaited++
+			}
 			if err != nil {
 				return false, err
 			}
 			i.permits = append(i.permits, permit)
 			curPermit = permit
 		} else {
 			// don't block if we are prefetching for a future seriesID.
-			permit, err := i.blockPermits.TryAcquire(ctx)
+			permit, acquireResult, err := i.blockPermits.TryAcquire(ctx)
```

Comment:
this is never going to wait.

in fact, this probably won't work because the algorithm will be greedy and prefetch as many series until it has to wait. so this will always error

Reply:
Made `TryAcquire` continue to return the `Permit` directly